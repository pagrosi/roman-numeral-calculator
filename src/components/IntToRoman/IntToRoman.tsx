import { useState } from 'react';
import { intRomanPairs } from '@/constants/intRomanPairs';
import { Input } from './Input';

export const convertIntToRoman = (int: number): string => {
	let result = '';
	let i = 0;

	while (int > 0) {
		if (int >= intRomanPairs[i].int) {
			result += intRomanPairs[i].roman;
			int -= intRomanPairs[i].int;
		} else {
			i += 1;
		}
	}

	return result;
};

export const IntToRoman = (): JSX.Element => {
	const [intInputNumber, setIntInputNumber] = useState(1);
	return (
		<>
			<Input value={intInputNumber} onChange={setIntInputNumber} />
			<div>
				Roman numerals output is:{' '}
				{convertIntToRoman(intInputNumber)}
			</div>
		</>
	);
};
