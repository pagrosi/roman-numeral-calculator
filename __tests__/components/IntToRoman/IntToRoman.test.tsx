import { render } from '@testing-library/react';
import { IntToRoman } from '@/components/IntToRoman';
import { convertIntToRoman } from '@/components/IntToRoman/IntToRoman';

describe('IntToRoman component', () => {
	test('renders an input element', () => {
		const { getByLabelText } = render(<IntToRoman />);
		const inputElement = getByLabelText('Integer input:');

		expect(inputElement).toBeInTheDocument();
	});

	test('renders correctly', () => {
		const { container } = render(<IntToRoman />);
		expect(container).toMatchSnapshot();
	});
});

describe('convertIntToRoman function', () => {
	test('returns empty string for zero', () => {
		expect(convertIntToRoman(0)).toBe('');
	});

	test('returns expected values for integers converted', () => {
		expect(convertIntToRoman(1)).toBe('I');
		expect(convertIntToRoman(50)).toBe('L');
		expect(convertIntToRoman(495)).toBe('CDXCV');
		expect(convertIntToRoman(1000)).toBe('M');
	});
});
