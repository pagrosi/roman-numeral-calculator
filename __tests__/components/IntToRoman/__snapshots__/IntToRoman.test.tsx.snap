// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntToRoman component renders correctly 1`] = `
<div>
  <label>
    Integer input:
     
    <input
      max="1000"
      min="1"
      type="number"
      value="1"
    />
  </label>
  <div>
    Roman numerals output is:
     
    I
  </div>
</div>
`;
